### DISCUSSION TITLE
Adding retry() logic when connecting to metadata DB

### DISCUSSION THREAD

**Discussion by Vishal2696** [apache/airflow#28302] [Q&A] [answered]
Hi Community, 

We are recently experiencing an issue with connecting to metadata DB and are currently thinking of a solution to avert this but not sure if it is possible and is the right way to solve it. So posting this for suggestions from community. 

 **Airflow version:** 2.2.3
 **Metadata DB:** Azure Database for PostgreSQL (V11) Connecting via pgbouncer running as a separate pod.
 **Deployment method:** Airflow on kubernetes using Celery Executor
 
 **Issue:** 
Airflow is experiencing connectivity issues with Metadata DB due to external networking factors. when this happens the worker is not able to fetch any information from the db (Variables, Connections etc..) Most of the time the issue is resolved automatically within few minutes however Airflow tries to connect to db once and if the attempt fails it is deemed as failure. 

**Our Proposed Solution:**
In the `execute()` method of our custom built plugin (This plugins gets values from Variables and Connections) adding an exception block to catch and check the error. If it's a "connection error with DB" then retry the same method after a short time interval. 

Is the above a right solution? 
_**Note:**_ _I'm aware that the permanent solution would be to investigate the network issues causing the failure in first place and I'm expecting the above only to be a temporary fix._

**Comment by potiuk** [comment, marked as answer]
Yes. This might work in your case

**Reply by Vishal2696** [reply]
Thank you! Gonna implement this and if it succeeds, will write a small article and post the link here.

**Comment by iw-pavan** [comment]
@Vishal2696 

Were you able to implement retries? Can you share code samples, It will be helpful

I am trying to implement similar to https://stackoverflow.com/questions/53287215/retry-failed-sqlalchemy-queries

But not able to override query class, @potiuk  Can you share details on how to achieve this.

PS: We are facing intermittent issues somewhere in connect

`File "/opt/app-root/lib64/python3.11/site-packages/psycopg2/__init__.py", line 122, in connect conn = _connect(dsn, connection_factory=connection_factory, **kwasync) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) connection to server at "<PostgresHost>, port 5432 failed: server closed the connection unexpectedly This probably means the server terminated abnormally before or while processing the request. (Background on this error at: https://sqlalche.me/e/14/e3q8)`
